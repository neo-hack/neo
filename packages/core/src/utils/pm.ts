import createStore, { PackageFilesResponse } from '@pnpm/package-store'
import createClient from '@pnpm/client'
import path from 'path'

import { CommonOptions } from '../interface'
import { STORE_PATH, NPM_REGISTRY } from './constants'

const authConfig = { registry: NPM_REGISTRY }

let pm: ReturnType<typeof createTemplatePM>

const createTemplatePM = async ({ storeDir = STORE_PATH }: CommonOptions) => {
  // @ts-ignore
  const { resolve, fetchers } = createClient.default({
    authConfig,
    cacheDir: path.join(STORE_PATH, '.cache'),
  })
  // @ts-ignore
  const storeController = await createStore.default(resolve, fetchers, {
    storeDir,
    verifyStoreIntegrity: true,
  })
  return {
    async request(alias: string, pref?: string) {
      const fetchResponse = await storeController.requestPackage(
        {
          alias,
          pref,
        },
        {
          downloadPriority: 0,
          lockfileDir: storeDir,
          preferredVersions: {},
          projectDir: storeDir,
          registry: NPM_REGISTRY,
          sideEffectsCache: false,
        },
      )
      return fetchResponse
    },
    async fetch() {
      // const pkgId = 'registry.npmjs.org/is-positive/1.0.0'
      const pkgId = 'registry.npmjs.org/ts-lib-template@latest'
      const fetchResponse = storeController.fetchPackage({
        force: false,
        lockfileDir: storeDir,
        pkg: {
          id: pkgId,
          resolution: {
            // integrity: 'sha1-iACYVrZKLx632LsBeUGEJK4EUss=',
            // registry: 'https://registry.npmjs.org/',
            // tarball: 'https://registry.npmjs.org/is-positive/-/is-positive-1.0.0.tgz',
            tarball: 'https://registry.npmjs.org/@aiou/ts-lib-template/-/ts-lib-template-0.5.0.tgz',
          },
        },
      })
      return fetchResponse
    },
    async import(to: string, response: PackageFilesResponse) {
      return storeController.importPackage(to, {
        filesResponse: response,
        force: false,
      })
    },
  }
}

const create = (params: CommonOptions) => {
  if (pm) {
    return pm
  }
  return createTemplatePM(params)
}

export default create
