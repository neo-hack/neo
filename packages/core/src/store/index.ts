import path from 'node:path'

import fs from 'fs-extra'
import tempy from 'tempy'

import { parseWantedPackage } from '../utils/find-pref-package'
import { debug } from '../utils/logger'
import createLockFile from './lock-file'
import createTemplatePM from './pm'

import type { CommonOptions } from '../interface'
import type { RequestOptions } from './pm'

type Options = RequestOptions & {
  name: string
  version?: string
}

const createStore = async (params: CommonOptions) => {
  debug.store('init store at %s', params.storeDir || 'default')
  const pm = await createTemplatePM(params)
  const lockFile = createLockFile(params)
  return {
    pm,
    lockFile,
    async add(params: { type?: 'preset' | 'template'; pref: string }) {
      const pkg = parseWantedPackage(params.pref)
      if (params.type === 'preset') {
        await this.addPreset({
          alias: pkg.alias,
          name: pkg.name,
          pref: params.pref,
          version: pkg.version,
        })
        return
      }
      await this.addTemplate({
        alias: pkg.alias,
        name: pkg.name,
        pref: params.pref,
        version: pkg.version,
      })
    },
    async addPreset(params: Options) {
      debug.store('add preset with params %O', params)
      const response = await pm.request({
        alias: params.alias,
        pref: params.pref,
        latest: params.latest,
      })
      const dir = tempy.directory()
      const importedPath = await pm.import(dir, response)
      if (!fs.existsSync(path.join(importedPath, 'index.json'))) {
        throw new Error('preset not found')
      }
      const pkgs = fs.readJsonSync(path.join(importedPath, 'index.json'))
      debug.store('preset templates list: %O', pkgs)
      // always update latest alias preset
      await lockFile.updatePreset({
        [params.name!]: pkgs,
      })
    },
    async addTemplate(params: Options) {
      debug.store('add template with params %O', params)
      const response = await pm.request({
        alias: params.alias,
        pref: params.pref,
        latest: params.latest,
      })
      if (!response) {
        debug.store('template not found')
        return
      }
      const { id } = response.body
      console.log('response.body', response.body)
      // templates pkgs maybe has same name, use id as key
      // name and pref parsed from user-input
      await lockFile.updateTemplates({
        [id]: {
          name: params.name,
          version: response.body.manifest?.version,
          id,
          pref: params.pref,
        },
      })
      return response
    },
  }
}

const create = async (params: CommonOptions) => {
  const store = await createStore(params)
  return store
}

export default create
