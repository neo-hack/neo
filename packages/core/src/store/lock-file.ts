import writeYamlFile from 'write-yaml-file'
import readYamlFile from 'read-yaml-file'
import path from 'path'
import fs from 'fs-extra'

import { LOCK_FILE, STORE_PATH } from '../utils/constants'
import { CommonOptions, LockFile, Package } from '../interface'
import { debugLogger } from '../utils/logger'

const getLockFilePath = (storeDir = STORE_PATH) => {
  return path.join(storeDir, LOCK_FILE)
}

let lockFile: ReturnType<typeof createLockFile>

const isCached = (cachedTemplates: Partial<Package>[], template: Partial<Package>) => {
  const { pref } = template
  return cachedTemplates.find((tpl) => {
    return tpl.pref === pref
  })
}

export const createLockFile = ({ lockFilePath }: { lockFilePath: string }) => {
  debugLogger.lockfile('lockfilepath at %s', lockFilePath)
  return {
    getLockFilePath() {
      return lockFilePath
    },
    async read(): Promise<LockFile> {
      // make sure lockfile
      if (!fs.existsSync(lockFilePath)) {
        return {}
      }
      return readYamlFile(lockFilePath)
    },
    async write(data: LockFile) {
      return writeYamlFile(lockFilePath, data)
    },
    async updatePreset(data: any) {
      const lockfile: LockFile = await this.read()
      debugLogger.lockfile('update preset %O', data)
      if (!lockfile.presets) {
        lockfile.presets = {}
      }
      lockfile.presets = { ...lockfile.presets, ...data }
      return this.write(lockfile)
    },
    async updateTemplates(data: any) {
      const lockfile: LockFile = await this.read()
      debugLogger.lockfile('update templates %O', data)
      if (!lockfile.templates) {
        lockfile.templates = {}
      }
      lockfile.templates = { ...lockfile.templates, ...data }
      return this.write(lockfile)
    },
    async readTemplates({ presetNames }: { presetNames?: string[] } = {}): Promise<
      Partial<Package>[]
    > {
      const lockFile = await this.read()
      const templates: LockFile['templates'] = lockFile.templates || {}
      const cachedTemplates = Object.values(templates).map((tpl) => ({ ...tpl, cached: true }))

      const presets: LockFile['presets'] = lockFile.presets || {}
      let presetTemplates: Partial<Package>[] = Object.keys(presets)
        .reduce((acc, cur) => {
          return acc.concat(
            presets[cur].templates.map((tpl) => ({ ...tpl, preset: cur, pref: tpl.name })),
          )
        }, [] as Partial<Package>[])
        .map((tpl) => ({ ...tpl, cached: !!isCached(cachedTemplates, tpl) }))
      debugLogger.lockfile('read templates with %O', presetNames)

      if (presetNames) {
        presetTemplates = presetTemplates.filter(
          (tpl) => !tpl.preset || presetNames.includes(tpl.preset),
        )
        return presetTemplates
      }
      const allTemplates = presetTemplates.concat(cachedTemplates)
      debugLogger.lockfile('templates list %O', allTemplates)
      return allTemplates
    },
  }
}

const create = (params?: CommonOptions) => {
  if (lockFile) {
    return lockFile
  }
  return createLockFile({ lockFilePath: getLockFilePath(params?.storeDir) })
}

export default create
