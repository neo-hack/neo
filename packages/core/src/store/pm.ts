import createStore, { PackageFilesResponse, PackageResponse } from '@pnpm/package-store'
import createClient from '@pnpm/client'
import path from 'path'

import { CommonOptions } from '../interface'
import { STORE_PATH, NPM_REGISTRY, CACHE_DIRNAME } from '../utils/constants'
import { debug } from '../utils/logger'

const authConfig = { registry: NPM_REGISTRY }
export type RequestOptions = {
  alias?: string
  pref?: string
  latest?: boolean
}

let pm: ReturnType<typeof createTemplatePM>

const createCtrl = async ({
  storeDir = STORE_PATH,
  offline,
}: CommonOptions & { offline: boolean }) => {
  // @ts-ignore
  const { resolve, fetchers } = createClient.default({
    authConfig,
    cacheDir: path.join(STORE_PATH, CACHE_DIRNAME),
    preferOffline: offline,
  })
  // @ts-ignore
  const storeCtrl = await createStore.default(resolve, fetchers, {
    storeDir,
    verifyStoreIntegrity: true,
  })
  return storeCtrl
}

export const createTemplatePM = async ({ storeDir = STORE_PATH }: CommonOptions) => {
  // @ts-ignore
  const storeCtrls = {
    offline: await createCtrl({ storeDir, offline: true }),
    online: await createCtrl({ storeDir, offline: false }),
  }
  return {
    async getCtrl(params: Pick<RequestOptions, 'latest'>) {
      return params.latest ? storeCtrls.online : storeCtrls.offline
    },
    async fetch({ alias, pref, latest = true }: RequestOptions): Promise<PackageResponse> {
      const storeCtrl = await this.getCtrl({ latest })
      const fetchResponse = await storeCtrl!.requestPackage(
        {
          alias,
          pref,
        },
        {
          downloadPriority: 0,
          lockfileDir: storeDir,
          preferredVersions: {},
          projectDir: storeDir,
          registry: NPM_REGISTRY,
          sideEffectsCache: false,
        },
      )
      return fetchResponse
    },
    async request({ alias, pref, latest }: RequestOptions): Promise<PackageResponse> {
      debug.pm(`request %s with %s`, alias, pref)
      let fetchResponse = await this.fetch({ alias, pref, latest }).catch((e) => console.error(e))
      if (!fetchResponse) {
        fetchResponse = await this.fetch({ pref: alias, latest })
      }
      debug.pm('request response.body %O', fetchResponse.body)
      return fetchResponse
    },
    async import(
      to: string,
      response?: PackageFilesResponse,
      params: Pick<RequestOptions, 'latest'> = { latest: false },
    ) {
      if (!response) {
        return
      }
      debug.pm('import template from store %s', response.fromStore)
      const storeCtrl = await this.getCtrl(params)
      return storeCtrl.importPackage(to, {
        filesResponse: response,
        force: false,
      })
    },
  }
}

const create = (params: CommonOptions) => {
  if (pm) {
    return pm
  }
  return createTemplatePM(params)
}

export default create
