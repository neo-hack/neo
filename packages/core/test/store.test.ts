import createStore from '../src/store'
import { createTemplatePM } from '../src/store/pm'
import {
  clearLockFile,
  randomStoreDir,
  readLockFile,
  storeDir as defaultStoreDir,
} from './helpers'

describe('pm', () => {
  afterEach(() => {
    clearLockFile()
  })
  it.todo('import github:neo-hack/actions-template should from store')
  it('request should work', async () => {
    const pm = await createTemplatePM({ storeDir: defaultStoreDir })
    const response = await pm.request({ alias: '@aiou/ts-lib-template' })
    expect(response).toBeDefined()
  })
  it('request with version should work', async () => {
    const pm = await createTemplatePM({ storeDir: defaultStoreDir })
    const response = await pm.request({ pref: '@aiou/bin-template@2.1.1' })
    expect(response.body.manifest?.version).toBe('2.1.1')
  })
})

describe('store', () => {
  // FIXME: latest option not working, will always install latest
  it('add template', async () => {
    const storeDir = randomStoreDir()
    const store = await createStore({ storeDir })
    await store.addTemplate({
      alias: '@aiou/bin-template',
      latest: true,
      name: 'bin',
      pref: '@aiou/bin-template',
    })
    expect(readLockFile(storeDir)).toMatchSnapshot()
  })
  it('add preset #debug', async () => {
    const storeDir = randomStoreDir()
    const store = await createStore({ storeDir })
    await store.addPreset({
      alias: '@aiou/preset-aiou',
      latest: true,
      name: 'aiou',
      pref: '@aiou/preset-aiou',
    })
    expect(readLockFile(storeDir)).toMatchSnapshot()
  })
  it('add(type=preset)', async () => {
    const storeDir = randomStoreDir()
    const store = await createStore({ storeDir })
    await store.add({
      type: 'preset',
      pref: '@aiou/preset-aiou',
    })
    expect(readLockFile(storeDir)).toMatchSnapshot()
  })
  it('add(type=template)', async () => {
    const storeDir = randomStoreDir()
    const store = await createStore({ storeDir })
    await store.add({
      type: 'template',
      pref: '@aiou/bin-template@3.0.1',
    })
    expect(readLockFile(storeDir)).toMatchSnapshot()
  })
})
