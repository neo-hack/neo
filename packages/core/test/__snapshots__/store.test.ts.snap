// Vitest Snapshot v1

exports[`store > add preset #debug 1`] = `"/Volumes/ORICO/Users/jiangwei/projects/neo/packages/core/.store/9de2a988-ff04-4571-85a5-0e9b753ddf35"`;

exports[`store > add preset 1`] = `
{
  "presets": {
    "aiou": {
      "configs": [
        {
          "name": "PNPM CI workflow",
          "pref": "./assets/workflows/pnpm/ci.yaml",
        },
        {
          "name": "PNPM Release workflow",
          "pref": "./assets/workflows/pnpm/release.yaml",
        },
        {
          "name": "PNPM Snapshot Release workflow",
          "pref": "./assets/workflows/pnpm/snapshot-release.yaml",
        },
        {
          "name": "Rust CI workflow",
          "pref": "./assets/workflows/rust/ci.yaml",
        },
        {
          "name": "Rust Homebrew Release workflow",
          "pref": "./assets/workflows/rust/release.homebrew.yaml",
        },
        {
          "name": "SWC plugin Release workflow",
          "pref": "./assets/workflows/rust/release.swc.yaml",
        },
        {
          "name": "SWC plugin Snapshot-Release workflow",
          "pref": "./assets/workflows/rust/snapshot-release.swc.yaml",
        },
      ],
      "templates": [
        {
          "name": "@aiou/ts-lib-template",
        },
        {
          "name": "@aiou/actions-template",
        },
        {
          "name": "@aiou/rollup-plugin-template",
        },
        {
          "name": "@aiou/rollup-template",
        },
        {
          "name": "@aiou/dayjs-plugin-template",
        },
        {
          "name": "@aiou/bin-template",
        },
        {
          "name": "@aiou/webext-template",
        },
        {
          "name": "@aiou/eslint-config-template",
        },
        {
          "name": "@aiou/webpack-loader-template",
        },
        {
          "name": "@aiou/react-template",
        },
        {
          "name": "@aiou/react-components-lib-template",
        },
        {
          "name": "@aiou/original-html-template",
        },
        {
          "name": "@aiou/chrome-extenstion-template",
        },
      ],
    },
  },
  "version": undefined,
}
`;

exports[`store > add template 1`] = `
{
  "templates": {
    "bin": {
      "id": "@aiou/bin-template/4.0.0",
      "name": "bin",
      "pref": "@aiou/bin-template",
      "version": "4.0.0",
    },
  },
  "version": undefined,
}
`;

exports[`store > add(type=preset) 1`] = `
{
  "presets": {
    "@aiou/preset-aiou": {
      "configs": [
        {
          "name": "PNPM CI workflow",
          "pref": "./assets/workflows/pnpm/ci.yaml",
        },
        {
          "name": "PNPM Release workflow",
          "pref": "./assets/workflows/pnpm/release.yaml",
        },
        {
          "name": "PNPM Snapshot Release workflow",
          "pref": "./assets/workflows/pnpm/snapshot-release.yaml",
        },
        {
          "name": "Rust CI workflow",
          "pref": "./assets/workflows/rust/ci.yaml",
        },
        {
          "name": "Rust Homebrew Release workflow",
          "pref": "./assets/workflows/rust/release.homebrew.yaml",
        },
        {
          "name": "SWC plugin Release workflow",
          "pref": "./assets/workflows/rust/release.swc.yaml",
        },
        {
          "name": "SWC plugin Snapshot-Release workflow",
          "pref": "./assets/workflows/rust/snapshot-release.swc.yaml",
        },
      ],
      "templates": [
        {
          "name": "@aiou/ts-lib-template",
        },
        {
          "name": "@aiou/actions-template",
        },
        {
          "name": "@aiou/rollup-plugin-template",
        },
        {
          "name": "@aiou/rollup-template",
        },
        {
          "name": "@aiou/dayjs-plugin-template",
        },
        {
          "name": "@aiou/bin-template",
        },
        {
          "name": "@aiou/webext-template",
        },
        {
          "name": "@aiou/eslint-config-template",
        },
        {
          "name": "@aiou/webpack-loader-template",
        },
        {
          "name": "@aiou/react-template",
        },
        {
          "name": "@aiou/react-components-lib-template",
        },
        {
          "name": "@aiou/original-html-template",
        },
        {
          "name": "@aiou/chrome-extenstion-template",
        },
      ],
    },
  },
  "version": undefined,
}
`;

exports[`store > add(type=template) 1`] = `
{
  "templates": {
    "@aiou/bin-template": {
      "id": "@aiou/bin-template/3.0.1",
      "name": "@aiou/bin-template",
      "pref": "@aiou/bin-template@3.0.1",
      "version": "3.0.1",
    },
  },
  "version": undefined,
}
`;
